docs/source/migrating_to_four.rst

```diff
@@ -0,0 +1,39 @@
+Migrating To PyCC v4.0.0
+========================
+
+Client Structure
+----------------
+
+Versions of PyCC greater than ``4.0.0`` are structured hierarchically
+such that ``CitrinationClient`` is a container for context specific
+subclients. At any point in the client hierarchy, you can see which
+methods and subclients are currently available by calling ``.__repr__()`` 
+on the current client.
+
+.. literalinclude:: code_samples/general/client_serialization.py
+
+The consequence of this is that all method invocations on the client must be directed to the appropriate subclient. In other words, rather than calling ``upload`` on the ``CitrinationClient`` instance, you must call it on the ``DataClient`` accessed by calling ``.data`` on the ``CitrinationClient``.
```

Comment:
You weren't able to attach methods to the CitrinationClient class?

Reply:
I was, see `citrination_client/client.py`. I need to update this documentation. But I also wonder if we should encourage the use of those meta-programmed methods. My intuition would say that they are nice to have around for at least partial backwards compatibility but that we should encourage direct access of the member clients.

Reply:
We should probably just discourage use of CitrinationClient completely in that case. Is that your idea?

Reply:
This is a pattern we use in the python client for the cmc api. FWIW, I've found it to be a pretty nice interface. It's a little confusing to have a billion top level methods - name spacing is kind of nice.

ymmv

Reply:
E.g., `client.data.find_by_property("band gap")` is more readable imo than `clint.find_by_property("band gap")`

Reply:
Or you just write good documentation :p 🤷‍♂️

Reply:
@sparadiso It was really there for backwards compatibility. It should be the case that people can upgrade to 4.x and, if all they have been doing is running queries, not need to make any changes to their scripts. We should consider removing the global client class completely in 5.x when that day comes.

I think that we want to follow a more boto3-like experience where you can create a [Session](http://boto3.readthedocs.io/en/latest/reference/core/session.html) that contains global configs like site URL and API key, then generate clients for particular applications from it.

Reply:
@kjaym it might in the future make more sense remove it entirely, but for right now I am going to leave it as an option for backwards compatibility.